addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -3311,7 +3311,7 @@ var SnippetsMenu = Widget.extend({
             elements: ".oe_snippet",
             scrollingElement: $scrollingElement[0],
             handle: '.oe_snippet_thumbnail:not(.o_we_already_dragging)',
-            cancel: '.oe_snippet.o_disabled',
+            ignore: '.oe_snippet.o_disabled, [data-module-id]',
```

Comment:
Shouldn't it be more "robust" in the way that it should not be retrieving the snippets itself but uses `this.$snippets` somehow, as before?
Right now, it feels like it's replicating what is done to generate `this.$snippets` but it's just correct until it's not (when we decide to exclude some other snippets from `this.$snippets` following a new IMP eg).
For tracking purpose, here is how `this.$snippets` is generated and why this diff is needed.
```js
 this.$snippets = $scroll.find('.o_panel_body').children()
            .addClass('oe_snippet')
            .each((i, el) => {
                // Some code
            })
            .not('[data-module-id]');
```

Reply:
Indeed, `ignore` seems to already accept a function from what I see :) `ignore: [String, Function],`

Reply:
> Indeed, ignore seems to already accept a function from what I see :) ignore: [String, Function],

It does accept a function, [but to compute a selector](https://github.com/odoo/odoo/blob/912a91817546e92a52df4e23bfd980d9ad4eb698/addons/web/static/src/core/utils/draggable_hook_builder.js#L683). 

So in the end, we would have something similar if we want to go that route. We could also use the computation of this.$snippets to just add a class to the elements we should ignore: `o_not_draggable` or something like that?

If we want to re-do what jQuery did (check if the element is part of the list of elements we want to drag), we would need to add a couple more hooks into the framework code, and it would be like re-coding jQueryUI.

I don't mind doing it, but I think this is more a case of "this makes less sense because it's in-between refactorings" and less a "the old way was better"

Reply:
Shouldn't `useDragAndDrop` take a NodeList as parameter, and not just a selector for instance?
Right now, it's just going to break at some point, it's obvious, as it's mimicking stuff instead of using the stuff itself.
If there is not better way, sure, why not, but it's sad when you know it **will** break and we just have to wait for it.

I didn't investigate the whole stack, but if it's a JS framework change that would be required to be able to do that, isn't it worth it, even in 17?

Reply:
> but I think this is more a case of "this makes less sense because it's in-between refactorings" and less a "the old way was better"

I don't understand what you mean @detrouxdev , can you clarify?

Reply:
I think I did not explain myself properly, had a quick call with RDE just to make it clear but I'll try to write it down:

It seems to me like the framework part is not made to take lists of elements, rather selectors. We could add code to support taking lists of elements. Or we could adapt our code to properly use the selectors.

In my opinion, the issue is not that the draggable tool cannot take a list of elements and more that my selector was not properly crafted to take into account that all "oe_snippets" might not be draggable. But since the API looks almost exactly like before, it looks like the mistake is that the tool cannot take the list of elements.

If our code was built thinking "the list of draggable elements are all elements which match `.oe_snippets:not(.o_disabled)`" and then add `o_disabled` to elements which we don't want to drag, we would not have an issue here. (Well still, we would have an issue cause of the wrong property used)

Because our code was built with this.$snippets, it makes sense to us to want to use it. But in a future refactoring maybe managing the list of snippets would not be the best strategy. (e.g. right now the component does not need to manage the list and make sure it's always up to date, so if we remove a snippet from the dom, it won't keep it in memory. Same if we add a new snippet to the dom, the drag and drop does not need to know, which could be an advantage). 

Obviously this all depends on how it will be implemented. But I'd rather not add code to the framework for something we might not use at all post OWL refactoring. If then we think it's a better idea to add it because it still makes sense then why not.

Instead, what I think we should do is, for now, add a class to `this.$snippets` just before starting the drag and drop options so that the selector can be properly crafted to only take `this.$snippets`.

In the future, I think instead of just excluding the "installable" elements from the list, we should take advantage of the `.o_disabled` class, and base the idea of snippets that are draggable on that. And if we want to disable more snippets, just make sure that `o_disabled` is set.

Hopefully that made sense 😄 

Still, if you think making the tool take a NodeList instead of a selector is better, I really don't mind doing it. I see advantages in both 👍🏻

Reply:
Just to clarify, I'm not trying to say I did not make a mistake, I did. I just think I miss-used the selectors instead of the tool not having the list of draggable elements.

Reply:
> I'm not trying to say I did not make a mistake, I did

Don't worry, no one is implying that ^^ It's just that QSM didn't know about our call, so it's nice that you summed it up.
And it's always easier to see the "mistakes" when reviewing than when coding, it's why there is a review ;-)

Reply:
Yep, of course, agreed with everything. Some notes though, in case creating a new class to properly couple the two behavior was not possible:
- I think in that case, temporarily changing the framework with a nice comment is better than relying on the fact things will be refactored in the future (even if we are 99.9999% sure OWL will be used, we may not have to change that, or it might be in 5 versions from now).
- If not that, then adding some sort of comment which allows for someone modifying the `this.$snippets` definition to find the place where we "duplicate" the selector. You'll find that I have added some `// grep: THINGS_I_WANT_TO_COUPLE` in the code :wink: